prov/tcp/src/tcpx_comm.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2017 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *	   Redistribution and use in source and binary forms, with or
+ *	   without modification, are permitted provided that the following
+ *	   conditions are met:
+ *
+ *		- Redistributions of source code must retain the above
+ *		  copyright notice, this list of conditions and the following
+ *		  disclaimer.
+ *
+ *		- Redistributions in binary form must reproduce the above
+ *		  copyright notice, this list of conditions and the following
+ *		  disclaimer in the documentation and/or other materials
+ *		  provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <rdma/fi_errno.h>
+#include <prov.h>
+#include <sys/types.h>
+#include <fi_util.h>
+
+#include "tcpx.h"
+
+int tcpx_cm_send(int fd, const void *buf, int len)
+{
+	int ret, done = 0;
+
+	while (done != len) {
+		ret = ofi_send_socket(fd, (const char*) buf + done,
+				      len - done, MSG_NOSIGNAL);
+		if (ret < 0) {
+			if (OFI_SOCK_TRY_SND_RCV_AGAIN(ofi_sockerr()))
+				continue;
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to write to fd: %s\n",
+				strerror(ofi_sockerr()));
+
+			return -FI_EIO;
+		}
+		done += ret;
+	}
+	return 0;
+}
+
+int tcpx_cm_recv(int fd, void *buf, int len)
+{
+	int ret, done = 0;
+	while (done != len) {
+		ret = ofi_recv_socket(fd, (char*) buf + done, len - done, 0);
+		if (ret <= 0) {
+			if (OFI_SOCK_TRY_SND_RCV_AGAIN(ofi_sockerr()))
+				continue;
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to read from fd: %s\n",
+				strerror(ofi_sockerr()));
+			return -FI_EIO;
+		}
+		done += ret;
+	}
+	return 0;
+}
```

Comment:
If the CM messages are wanting blocking socket semantics, I would make/leave the socket as blocking until the CM exchange completes.  Then ofi_send_socket() can be used as-is, and ofi_recv_socket() can be called using MSG_WAITALL.

Reply:
Can I assume that connection data is small enough to not cause hangs if we use MSG_WAITALL flag?

Reply:
Well, this function is essentially blocking anyway.  Btw, as an optimization, you could post the send using sendmsg(), so that the header + cm data go out as a single packet.  A TCP connection _should_ start with 1 mtu worth of available window space.

Reply:
I see calls further down that set the sockets to nonblocking after the CM messages are exchanged.  If that's the case, then we should be able to remove these functions and call ofi_recv_socket/send directly.